File: app/src/main/java/com/testdemo/testCanDragLayout/DraggableLayout.java
package com.testdemo.testCanDragLayout;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.animation.AnticipateInterpolator;
import android.view.animation.BounceInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.LinearLayout;

/**
 * Created by Administrator on 2018/2/9.
 */
public class DraggableLayout extends LinearLayout {
    private final static String TAG = "greyson_dragLayout";

    private float lastTouchY;
    private float firstTouchY;


    public DraggableLayout(Context context) {
        super(context);
    }

    public DraggableLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public DraggableLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        float y = ev.getY();
        switch (ev.getAction()) {
            case MotionEvent.ACTION_DOWN:
                Log.i(TAG, "event down y = " + getY());
                lastTouchY = y;
                firstTouchY = y;
                break;

            case MotionEvent.ACTION_MOVE:
                if (lastTouchY - y != 0) {
                    return true;
                }
                break;
        }
        return super.onInterceptTouchEvent(ev);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float y = event.getY();
        switch (event.getAction()) {

            case MotionEvent.ACTION_MOVE:
                float translationOffset = y - firstTouchY;//滑动的总偏移量
                float currentY = (float) Math.sqrt(Math.abs(translationOffset)) * 2;
                if (y - firstTouchY < 0) {
                    setTranslationY(-currentY);
                } else {
                    setTranslationY(currentY);
                }
//                Log.i(TAG, "offsetSmall=" + offsetSmall + "  offset = " + offset + "  top=" + getTop());
                lastTouchY = y;
                break;

            case MotionEvent.ACTION_UP:
                Log.i(TAG, "event up y = " + getY() + "  translationY=" + getTranslationY());
                ObjectAnimator objectAnimator = ObjectAnimator.ofFloat(this, "translationY", 0f);
                objectAnimator.setDuration(300);
//                objectAnimator.setInterpolator(new BounceInterpolator());
                objectAnimator.setInterpolator(new DecelerateInterpolator());
                objectAnimator.start();
                break;
        }
        return true;
    }
}

File: app/src/main/java/com/testdemo/testCanDragLayout/DraggableLayout.kt
package com.testdemo.testCanDragLayout

import android.animation.ObjectAnimator
import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.animation.DecelerateInterpolator
import android.widget.LinearLayout

class DraggableLayout : LinearLayout {
    private val TAG = "greyson_dragLayout"

    private var lastTouchY = 0f
    private var firstTouchY = 0f

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun onInterceptTouchEvent(ev: MotionEvent?): Boolean {
        ev?.let {
            val y = it.y
            when (it.action) {
                MotionEvent.ACTION_DOWN -> {
                    Log.i(TAG, "event down y = $y")
                    lastTouchY = y
                    firstTouchY = y
                }

                MotionEvent.ACTION_MOVE -> return lastTouchY != y
            }
        }
        return super.onInterceptTouchEvent(ev)
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        event?.let {
            val y = it.y
            when (it.action) {
                MotionEvent.ACTION_MOVE -> {
                    val translationOffset = y - firstTouchY//滑动的总偏移量
                    val currentY = (Math.sqrt(Math.abs(translationOffset).toDouble()) * 2).toFloat()
                    when (y - firstTouchY < 0) {
                        true -> translationY = -currentY
                        false -> translationY = currentY
                    }
                    lastTouchY = y
                }

                MotionEvent.ACTION_UP -> {
                    Log.i(TAG, "event up y = $y  translationY=$translationY")
                    val objectAnimator = ObjectAnimator.ofFloat(this, "translationY", 0f)
                    objectAnimator.duration = 300
                    objectAnimator.interpolator = DecelerateInterpolator()
                    objectAnimator.start()
                }
            }
        }
        return true
    }
}